### PR TITLE
CORDA-1578: Add flow state machine extension tutorial

### PR DESCRIPTION
**Pull request by exFalso**

### CODE REVIEW

File: docs/source/contributing-flow-state-machines.rst [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+.. container:: codeset
+
+    .. literalinclude:: ../../docs/source/example-code/src/main/kotlin/net/corda/docs/tutorial/flowstatemachines/TutorialFlowAsyncOperation.kt
+        :language: kotlin
+        :start-after: DOCSTART ExampleSummingFlow
+        :end-before: DOCEND ExampleSummingFlow
+
+That's it! Obviously this is a mostly useless example, but this is the basic code structure one could extend for heavier
+computations/other IO. For example the function could call into a ``CordaService`` or something similar. One thing to
+note is that the operation executed in ``execute`` must be redoable(= "idempotent") in case the node fails while the
```
Comment (tlil) [inline docs/source/contributing-flow-state-machines.rst]: `while`?

File: docs/source/contributing-flow-state-machines.rst [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+there's any.
+
+In this case the work involves persisting a checkpoint together with some deduplication data in a database transaction,
+then acknowledging any triggering messages, signalling that the flow has started, and creating a fresh database
+transaction, to be used by user code.
+
+The continuation is a ``Resume``, which instructs the state machine to hand control to user code. The state change is
+a simple update of bookkeeping data.
+
+On other words the first transition concerns initializing of the flow, which includes the creation of the
```
Comment (tlil) [inline docs/source/contributing-flow-state-machines.rst]: `In other words`

Comment (tlil) [reply docs/source/contributing-flow-state-machines.rst]: `the initialization of the flow`

File: docs/source/contributing-flow-state-machines.rst [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+checkpoint.
+
+The next transition is the suspension on our summing operation, triggered by the ``Suspend`` event. As we can see in
```
Comment (tlil) [inline docs/source/contributing-flow-state-machines.rst]: `suspension of our`

File: docs/source/contributing-flow-state-machines.rst [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+this transition we aren't doing any work related to the summation yet, we're merely persisting the checkpoint that
+indicates that we want to do the summation. Had we added a ``toString`` method to our ``SummingOperationThrowing`` we
+would see a nicer message.
+
+The next transition is the faulty one, as we can see it was also triggered by a ``DoRemainingWork``, and executed our
+operation. We can see that there are two state "diff"s printed, one that would've happened had the transition succeeded,
+and one that actually happened, this is the marking of the flow's state as errored. The rest of the transitions involve
```
Comment (tlil) [inline docs/source/contributing-flow-state-machines.rst]: Rephrase `this is the marking of the flow's state as errored` ?

File: docs/source/contributing-flow-state-machines.rst [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+we'll store an instance of this class whenever we're suspending on such an operation. If the node fails or restarts
+while the operation is underway this class will be deserialized from the checkpoint and ``execute`` will be called
+again.
+
+Now we can use the internal function ``executeAsync`` to execute this operation from a flow.
+
+.. container:: codeset
+
+    .. literalinclude:: ../../core/src/main/kotlin/net/corda/core/internal/FlowAsyncOperation.kt
+        :language: kotlin
```
[the remark this answers is not in the document]
Comment (exFalso) [reply docs/source/contributing-flow-state-machines.rst]: Yes, the highlighter doesn't cover the complete Kotlin syntax

**Comment by exFalso** [comment]
@tlil addressed comments

File: docs/source/contributing-flow-state-machines.rst [hunk 6]
```diff
@@ -0,0 +1,237 @@
...
+        :start-after: DOCSTART FlowAsyncOperation
+        :end-before: DOCEND FlowAsyncOperation
+
+Let's imagine we want to add a suspending operation that takes two integers and returns their sum. To do this we
+implement ``FlowAsyncOperation``:
+
+.. container:: codeset
+
+    .. literalinclude:: ../../docs/source/example-code/src/main/kotlin/net/corda/docs/tutorial/flowstatemachines/TutorialFlowAsyncOperation.kt
+        :language: kotlin
```
[the remark this answers is not in the document]
Comment (tlil) [reply docs/source/contributing-flow-state-machines.rst]: @exFalso mind adding a Java sample as well?

**Comment by exFalso** [comment]
@tlil added Java samples

**Comment by tlil** [comment]
:+1: Looks good to me. The `Participate` section is a bit confusing at the moment, so I will file a follow-up ticket and get that sorted.

**Comment by mikehearn** [comment]
Nice!!

**Pull request merged**